README.rst
=====
sensu
=====

Install and configure sensu and some simple common checks (disk, memory etc.)

.. note::

    See the full `Salt Formulas installation and usage instructions
    <http://docs.saltstack.com/topics/development/conventions/formulas.html>`_.

Dependencies
============

Playbooks for checks are available in `docs/playbooks <https://github.com/ministryofjustice/sensu-formula/tree/master/docs/playbooks/>`_.

Dependencies
============

.. note::

   This formula has a dependency on the following salt formulas:

   `logstash <https://github.com/ministryofjustice/logstash-formula>`_

   `redis <https://github.com/ministryofjustice/redis-formula>`_

   `rabbitmq <https://github.com/ministryofjustice/rabbitmq-formula>`_

   `nginx <https://github.com/ministryofjustice/nginx-formula>`_


Adding checks via pillar data
=============================

All checks are now defined in the map.jinja or in the pillar (which overrides the map.jinja).

They can be of type 'graphite', 'procs', or 'basic': mapping to the macros sensu_check_graphite,
sensu_check_procs, and sensu_check respectively.

To do so, add a pillar under 'sensu' called check_definitions::

  sensu:
    check_definitions:
      carbon_queue_size:
        type: graphite
        target: 'sumSeries(carbon.agents.*.cache.size)'
        description: 'Graphite Carbon Cache Size'
        playbook: 'https://gist.github.com/mikepea/fda477167608c2b93512'
        subscribers:
          - monitoring.server
      logstash_redis_queue_size:
        type: basic
        command: '/etc/sensu/community/plugins/redis/check-redis-list-length.rb -k logstash:beaver -w 1000 -c 100000'
        description: 'Logstash queue size'
        playbook: 'https://gist.github.com/mikepea/dad0d07ac012cfe45855'
        subscribers:
          - monitoring.server
      monitoring_server_redis_memory_size:
        type: basic
        command: '/etc/sensu/community/plugins/redis/check-redis-memory.rb -w 20000 -c 80000'
        description: 'Monitoring Server Redis memory'
        playbook: 'https://gist.github.com/mikepea/dad0d07ac012cfe45855'
        subscribers:
          - monitoring.server
      exit_0_test:
        type: basic
        command: 'exit 0'
        interval: 10
        playbook: 'https://gist.github.com/mikepea/648a1e8877cf0c9955a9'
      postgres:
        type: procs
        subscribers:
          - monitoring.server
        playbook: 'https://gist.github.com/mikepea/695d1d5d08b9106d9cf2'
      udevd:
        type: procs
        playbook: 'https://gist.github.com/mikepea/1ada349535df0f2a6cb6'

Setting handlers in pillar
==========================

See below for enabling handlers. Once enabled you can specify a handler explicitly in the check definition::

    sensu:
      check_definitions:
        used-root-disk:
          handlers:
            - hipchat
            - pagerduty

Available states and macros
===========================

.. contents::
    :local:

``server``
----------

Install sensu server, dashboard and api components talking against a local
rabbitmq cluster and local redis server

Will install redis and rabbitmq servers, and assert the rabbitmq vhost and
user.

Example usage::

    include:
      - sensu.server

``client``
----------

Install sensu client and configure it to connect to the sensu server.

The client will be subscribe to checks on the 'all' channel, and to everything
in the ``roles`` grain.

Example usage::

    include:
      - sensu.client

Pillar variables
~~~~~~~~~~~~~~~~

The client will connect to the sensu server via rabbit MQ, controlled by the
following pillar values. It will default to connecting to monitoring.local on
the default rabbitmq port.

- sensu:rabbitmq:host

- sensu:rabbitmq:port

- sensu:rabbitmq:vhost

- sensu:rabbitmq:user

- sensu:rabbitmq:password

``sensu_check`` macro
---------------------

Macro to create a new check instance.

The macro has the following arguments:

name
  A for the check name. Must be unique on the enviornment

command
  The command to run for the check

handlers
  The type of the entries in this log file. Shows up as the type field in
  logstash.

  **Default:** [``default``]

interval
  How often (in seconds) to run the check

  **Default:** 60

subscribers
  Which clients should perform this check

  **Default:** [``all``]

occurrences
  Number of failure occurrences before the handler should take action

  **Default:** 1

playbook
  URL of a doc explaining how to deal with this alert. This will be used for
  hipchat notifier and possibly other handler types.

Example usage::

    include:
      - sensu.server

    {% from 'sensu/lib.sls' import sensu_check with context %}
    {# This check is included by default #}
    {{ sensu_check('check_swap', '/etc/sensu/plugins/system/check-swap-percentage.sh -w 5 -c 25') }}
    {# This check is better done as the sensu_check_proc macro though#}
    {{ sensu_check('check_swap', '/etc/sensu/community/plugins/processes/check-procs.rb -p salt-master -C 1', subscribers=['master'] }}



``sensu_check_graphite`` macro
------------------------------

Macro to perform a check against a graphite metric target

The macro accepts the following arguments in addition to those of the ```sensu_check`` macro`_:

name
  A for the check name. Must be unique on the box

metric_name
  The name of the metric/target to pull from graphite. This can be any standard graphite target
  and can therefore include any of the default graphite functions. If the test is host-specific
  the test can also refer to the hostpath by using the ``:::metric_prefix:::`` sensu variable.

desc
  The description of the check. This is used when generating alerts.

params
  The set of additional command line parameters for this check. This should
  either include the warning and critical levels, or the levels must be defined
  in the pillar - but not both.  For more details on the available options
  please consult the graphite check at
  ``./sensu/files/plugins/graphite-data.rb``.


Configuring thresholds
~~~~~~~~~~~~~~~~~~~~~~

This macro will look in the pillar under ``sensu:checks`` for a dictionary that
matches the check name (``free-root-disk`` in this example) and if that
contains ``warning`` or ``critical`` keys it will use those values and append
``-w`` and ``-c`` options to the params automatically.

Example usage::

    include:
      - sensu.server

    {% from 'sensu/lib.sls' import sensu_check_graphite with context %}
    {{ sensu_check_graphite("free-root-disk",
                        "metrics.:::metric_prefix:::.df.root.df_complex.free",
                        "--below -a 600",
                        "Root Disk Full") }}

With the following pillar (which is the default)::

    sensu:
      checks:
        free-root-disk:
            warning: 10737418240
            critical: 5368709120


``sensu_check_procs`` macro
---------------------------

Install a sensu check to make sure that the named process exists

The macro has the following arguments in addition to those of the
```sensu_check`` macro`_:

name
  The process name to check for.

  This will form a sensu check named 'process-' + ``name``

pattern
  If the pattern you want to check for is not 'url' safe then you can
  explicitly specify pattern to look for.

  For example if you want to check for ``mongod`` but not ``mongodump`` then
  you would specify a pattern of ``mongod$``

  **Default:** the same value as the name parameter

critical_under
  Raise an critical alert when there are fewer than this many processes matched

  **Default:** ``1``

critical_over
  Raise an critical alert when there are greater than this many processes
  matched

Example usage::

    include:
      - sensu.server

    {% from 'sensu/lib.sls' import sensu_check_procs with context %}
    {{ sensu_check_procs("salt-master", subscribers=["master"]) }}
    {{ sensu_check_procs("mongod", pattern="mongod$") }}



Notifications
=============

By default the sensu server will only generate notifications to STDOUT and therefore they will only be
visible in the dashboard and in sensu-server.log. To enable additional notification methods you need to 
enable them in the pillar. You can enable as many as you like of the additional notifications.

Email
-----

Basic email notification.

Example::

    sensu:
      notify:
        email: 'alerts@mydomain.com'


Mailer
------

Advanced email notification. Provides more insight into the issue then basic email notification.

The mailer handler handler sends formatted emails via the configured SMTP server.  With the default settings
a local MTA is required.

Example::

    sensu:
      notify:
        mailer_mail_to: 'user@host.com'
        mailer_mail_from: 'sensu@sensu.local'
        mailer_smtp_address: 'localhost'
        mailer_smtp_port: '25'
        mailer_smtp_domain: 'sensu.local'
        mailer_admin_gui: 'http://sensu.local'


HipChat
-------

You need to obtain an APIkey from Hipchat Admin. By default, if a roomname isn't specified it will sent Alerts
to the 'Alerts' room.

Or you can use your own api key if you bump the apiversion to v2 (it defaults to v1).

Example::

    sensu:
      notify:
        hipchat_apikey: c5wzTko0O59Xb6wlIKRstaQLbcsJJJFAANaEoD3
        hipchat_roomname: 'My Project Alerts'
        hipchat_apiversion: v1

Multiple room notifications can be enabled by passing an array of room names such as ["Room1", "Room2"]

Pagerduty
---------

To integrate with Pagerduty, you must first create a Service definition which is driven by an API key. 
Once you have this, you should add the generated API key to the default pillar.

Example::

    sensu:
      notify:
        pagerduty_apikey: 9e880a23f5ab1103bb7279896804e8a0

You must also activate this handler for each check.

Example::

    sensu:
      check_definitions:
        used-root-disk:
          handlers:
            - pagerduty

Grafana Integration
-------------------

We have customised the hipchat.rb community handler to also integrate with
Grafana.

To enable this, add the following pillar setting::

    sensu:
      notify:
        hipchat_grafana_base: '{{grafana_http_path}/dashboard/script/custom_metrics.js?no_help&'

Any metric based around the 'sensu_check_graphite' macro should then
automatically get a link to the respective data in Grafana.

apparmor
========

This formula includes profiles for all the sensu components. Apparmor is by
default in complain mode which means it allows the action and logs. To make it
deny actions that the beaver profile doesn't cover set the following pillar::

    apparmor:
      profiles:
        sensu_api:
          enforce: ''
        sensu_client
          encorce: ''
        sensu_dashboard:
          encorce: ''
        sensu_server:
          encorce: ''


unix-socket-backlog
===================

This check will find connections to a unix socket that are still connecting.

Example::

    {% from 'sensu/lib.sls' import sensu_check with context %}
    {{ sensu_check('unix-socket-backlog', '/etc/sensu/plugins/unix-socket-backlog.rb -s /var/run/unicorn.sock -w 1 -c 5', subscribers=['www']) }}



CHANGELOG.md
## UNRELEASED

* Updated to final sensu-community-plugins commit

## Version 5.2.4

* Fix `occurrences` for sensu_check_graphite not ending up in the file on disk

## Version 5.2.3

* Add a --pretty option to the graphite plugin for simpler messages
* Check the ubuntu version before enabling ptrace for apparmor

## Version 5.2.2

* Fixed bug in the URL assembly for hipchat messages. 
  The Graph link in the hipchat message was broken.

* Increased alert threshold for NTP offset empty datasets 
  from 20 mins to 6hr, after getting spurious noisy 
  intermittent alerts on PVB.

## Version 5.2.1

* Fix sensu handlers.json template to allow multiple hipchat rooms using the 
  correct json syntax

## Version 5.2.0

* Increased default ntp offset alert threshold from 10 to 20 mins, to reduce
  redundant alert noise
* Extend hipchat handler to be able to send alerts to multiple rooms at once

  Simple make sensue:notify:hipchat_roomname an array and it will send it to
  all those rooms

## Version 5.1.4

* Add `ps` and `sh` to sensu-client apparmor profile to silence normal
  behavior

## Version 5.1.3

* Restart sensu-api as well as sensu-server when checks change

## Version 5.1.2

* Restart sensu-server when checks are removed as well as added

## Version 5.1.1

* Add a collectd script to graph redis process information (memory, cpu, etc)

## Version 5.1.0

* Extend check-elastic.rb to run count/threshold based checks too.

## Version 5.0.1

* Fix syntax error in handlers.json

## Version 5.0.0

* Remove pagerduty from default handlers
* Upgrade Uchiwa to 0.7.0-1
* Move all checks to map.jinja, this makes them configurable from the pillar.

## Version 4.5.2

* Move disk-space check to map.jinja so that it is overrideable in pillar.

## Version 4.5.1

* Add a new handler for sending selected alerts to the 2nd line support dashboard.

## Version 4.5.0

* Add 'Graph' link to integrate Grafana with hipchat notifications
* Can now have graphite checks which check the value is in a min/max range.

## Version 4.4.4

* Remove supervisor dependency
* Add ability to use the mailer handler from the community notification plugins.

## Version 4.4.3

* Sets the correct permissions on the plugins, so that only sensu can execute.

## Version 4.4.2

* Fix uchiwa pin for Trusty

## Version 4.4.1

* Sensu logging level is configurable
* Fix uchiwa order

## Version 4.4.0

* Sensu checks can now be provided by pillar data.
* Playbook for apparmor check, resolves #42
* Fix to allow standalone checks to be used again
* Allow sensu-embedded gems to be installed
* Silence apparmor message about 'sh'

## Version 4.3.0

* Upgrade Uchiwa to 0.3 (golang version)
* Update nginx conf to allow websockets over proxy.

## Version 4.2.0

* Prevent graphite-data.rb from throwing NilClass error on bad target/no data
* Uchiwa locked at 0.2.4 due to Websockets issue, but code present to upgrade to 0.3
* Upgrade sensu to 0.13.1-1

## Version 4.1.1

* Pin uchiwa for Ubuntu 12.04

## Version 4.1.0

* Let us specify a more exact process pattern to check for
* Allow passing through subscribers, handlers and all other parameters from
  sensu_check_graphite and sensu_check_proc down to the base sensu_check
* Improve docs

## Version 4.0.0

* **BREAKING** changes, see UPGRADING.md Move to sensu version 0.13.1 and uchiwa dashboard.

## Version 3.3.1

* Allow defining subscribers on graphite checks.

## Version 3.3.0

* Fixes to ensure sensu-server/sensu-client restarted on update
* Add swap-out graphite check
* Increase thesholds for swap-used check to reflect addition of swap-out
* Add Playbooks for swap-used and swap-out checks
* Add an alert on ntp clock drift to make sure time of all boxes is accurate

## Version 3.2.2

* Pagerduty should only send critical alerts
* Add occurences to check macro, set graphite checks to 2 occurrences.

## Version 3.2.1

* Fix apparmor profile for various new checks.
* Change apparmor to use generic elastic search check

## Version 3.2.0

* New plugin to check elastic search for user specified query.

  This lets us raise alerts based on log lines. See help in
  sensu/files/plugins/check-elastic.rb for usage

* Fix apparmor profile for sensu-server
* Add pillar entry to change hipchat api version used in sending notifications

## Version 3.1.1

* Pin sensu version for old dashboard service to work. (temp fix)

## Version 3.1.0

* Check JSON definitions only get installed on the server, no longer the
  client.
* Set subscriptions based on salt roles. Note your roles must be in a list!
* Included a new plugin to check backlog on unix sockets.

## Version 3.0.1

* Fix RabbitMQ start ordering

## Version 3.0.0

* Replace absolute disk space free check with one that checks percent in use
* Replace absolute swap check with one that checks percent used.
* Add apparmor profiles for sensu processes (client,api,dashboard and server)

## Version 2.2.0

* Allow graphite host to be specified in the pillar
* Allow check limits for graphite checks to be specified from pillar values

## Version 2.1.5

* Fix typo introduced in v2.1.4

## Version 2.1.4

* Be explicit about the permissions of the files we managed.

  Without this you could end up with a broken configuration depending on the
  umask of the user you run `salt-call` as.

## Version 2.1.3

* Ensure that /etc/salt/conf.d/checks is managed idempotently. Incremental
  runs of highstate will not incorrectly mark all files as changed.
* Pin community-plugins to a git commit instead of just master. This can
  be overridden by pillar var defined in map.jinja

## Version 2.1.2

* Fix: redphone and hipchat gems are now correctly installed in embedded
  ruby when pagerduty/hipchat integration is enabled
* Changed the Hipchat API to v1 to avoid the API key being associated
  with a person
* Removed additional subscriptions based on roles. This is not used at
  present and causes a failure on machines where roles are not defined

## Version 2.1.1


_states/sensu.py
# -*- coding: utf-8 -*-

# Import python libs
import os
import shutil
import difflib
import logging
import re
import fnmatch
import json
import pprint
import traceback

# Import third party libs
import yaml

# Import salt libs
import salt.utils
import salt.utils.templates
from salt.exceptions import CommandExecutionError
from salt._compat import string_types, integer_types

log = logging.getLogger(__name__)

COMMENT_REGEX = r'^([[:space:]]*){0}[[:space:]]?'

_ACCUMULATORS = {}
_ACCUMULATORS_DEPS = {}


def _error(ret, err_msg):
    ret['result'] = False
    ret['comment'] = err_msg
    return ret


def _test_owner(kwargs, user=None):
    '''
    Convert owner to user, since other config management tools use owner,
    no need to punish people coming from other systems.
    PLEASE DO NOT DOCUMENT THIS! WE USE USER, NOT OWNER!!!!
    '''
    if user:
        return user
    if 'owner' in kwargs:
        log.warning(
            'Use of argument owner found, "owner" is invalid, please '
            'use "user"'
        )
        return kwargs['owner']

    return user

def _check_user(user, group):
    '''
    Checks if the named user and group are present on the minion
    '''
    err = ''
    if user:
        uid = __salt__['file.user_to_uid'](user)
        if uid == '':
            err += 'User {0} is not available '.format(user)
    if group:
        gid = __salt__['file.group_to_gid'](group)
        if gid == '':
            err += 'Group {0} is not available'.format(group)
    return err


def _managed(name,
            source=None,
            source_hash='',
            user=None,
            group=None,
            mode=None,
            template=None,
            makedirs=False,
            dir_mode=None,
            context=None,
            replace=True,
            defaults=None,
            env=None,
            backup='',
            show_diff=True,
            create=True,
            contents=None,
            contents_pillar=None,
            **kwargs):
    '''
    Manage a given file, this function allows for a file to be downloaded from
    the salt master and potentially run through a templating system.

    name
        The location of the file to manage

    source
        The source file to download to the minion, this source file can be
        hosted on either the salt master server, or on an HTTP or FTP server.
        For files hosted on the salt file server, if the file is located on
        the master in the directory named spam, and is called eggs, the source
        string is salt://spam/eggs. If source is left blank or None
        (use ~ in YAML), the file will be created as an empty file and
        the content will not be managed

        If the file is hosted on a HTTP or FTP server then the source_hash
        argument is also required

    source_hash
        This can be one of the following:
            1. a source hash string
            2. the URI of a file that contains source hash strings

        The function accepts the first encountered long unbroken alphanumeric
        string of correct length as a valid hash, in order from most secure to
        least secure::

            Type    Length
            ======  ======
            sha512     128
            sha384      96
            sha256      64
            sha224      56
            sha1        40
            md5         32

        The file can contain several checksums for several files. Each line
        must contain both the file name and the hash.  If no file name is
        matched, the first hash encountered will be used, otherwise the most
        secure hash with the correct source file name will be used.

        Debian file type ``*.dsc`` is supported.

        Examples::

            /etc/rc.conf ef6e82e4006dee563d98ada2a2a80a27
            sha254c8525aee419eb649f0233be91c151178b30f0dff8ebbdcc8de71b1d5c8bcc06a  /etc/resolv.conf
            ead48423703509d37c4a90e6a0d53e143b6fc268

        Known issues:
            If the remote server URL has the hash file as an apparent
            sub-directory of the source file, the module will discover that it
            has already cached a directory where a file should be cached. For
            example:

            .. code-block:: yaml

                tomdroid-src-0.7.3.tar.gz:
                  file.managed:
                    - name: /tmp/tomdroid-src-0.7.3.tar.gz
                    - source: https://launchpad.net/tomdroid/beta/0.7.3/+download/tomdroid-src-0.7.3.tar.gz
                    - source_hash: https://launchpad.net/tomdroid/beta/0.7.3/+download/tomdroid-src-0.7.3.tar.gz/+md5


    user
        The user to own the file, this defaults to the user salt is running as
        on the minion

    group
        The group ownership set for the file, this defaults to the group salt
        is running as on the minion

    mode
        The permissions to set on this file, aka 644, 0775, 4664

    template
        If this setting is applied then the named templating engine will be
        used to render the downloaded file, currently jinja, mako, and wempy
        are supported

    makedirs
        If the file is located in a path without a parent directory, then
        the state will fail. If makedirs is set to True, then the parent
        directories will be created to facilitate the creation of the named
        file.

    dir_mode
        If directories are to be created, passing this option specifies the
        permissions for those directories. If this is not set, directories
        will be assigned permissions from the 'mode' argument.

    replace
        If this file should be replaced.  If false, this command will
        not overwrite file contents but will enforce permissions if the file
        exists already.  Default is True.

    context
        Overrides default context variables passed to the template.

    defaults
        Default context passed to the template.

    backup
        Overrides the default backup mode for this specific file.

    show_diff
        If set to False, the diff will not be shown.

    create
        Default is True, if create is set to False then the file will only be
        managed if the file already exists on the system.

    contents
        Default is None.  If specified, will use the given string as the
        contents of the file.  Should not be used in conjunction with a source
        file of any kind.  Ignores hashes and does not use a templating engine.

    contents_pillar
        .. versionadded:: 0.17.0

        Operates like ``contents``, but draws from a value stored in pillar,
        using the pillar path syntax used in :mod:`pillar.get
        <salt.modules.pillar.get>`. This is useful when the pillar value
        contains newlines, as referencing a pillar variable using a jinja/mako
        template can result in YAML formatting issues due to the newlines
        causing indentation mismatches.

        For example, the following could be used to deploy an SSH private key:

        .. code-block:: yaml

            /home/deployer/.ssh/id_rsa:
              file.managed:
                - user: deployer
                - group: deployer
                - mode: 600
                - contents_pillar: userdata:deployer:id_rsa

        This would populate ``/home/deployer/.ssh/id_rsa`` with the contents of
        ``pillar['userdata']['deployer']['id_rsa']``. An example of this pillar
        setup would be like so:

        .. code-block:: yaml:

            userdata:
              deployer:
                id_rsa: |
                  -----BEGIN RSA PRIVATE KEY-----
                  MIIEowIBAAKCAQEAoQiwO3JhBquPAalQF9qP1lLZNXVjYMIswrMe2HcWUVBgh+vY
                  U7sCwx/dH6+VvNwmCoqmNnP+8gTPKGl1vgAObJAnMT623dMXjVKwnEagZPRJIxDy
                  B/HaAre9euNiY3LvIzBTWRSeMfT+rWvIKVBpvwlgGrfgz70m0pqxu+UyFbAGLin+
                  GpxzZAMaFpZw4sSbIlRuissXZj/sHpQb8p9M5IeO4Z3rjkCP1cxI
                  -----END RSA PRIVATE KEY-----

        .. note::

            The private key above is shortened to keep the example brief, but
            shows how to do multiline string in YAML. The key is followed by a
            pipe character, and the mutli-line string is indented two more
            spaces.
    '''
    # Make sure that leading zeros stripped by YAML loader are added back
    mode = __salt__['config.manage_mode'](mode)

    user = _test_owner(kwargs, user=user)
    ret = {'changes': {},
           'comment': '',
           'name': name,
           'result': True}
    if not create:
        if not os.path.isfile(name):
            # Don't create a file that is not already present
            ret['comment'] = ('File {0} is not present and is not set for '
                              'creation').format(name)
            return ret
    u_check = _check_user(user, group)
    if u_check:
        # The specified user or group do not exist
        return _error(ret, u_check)
    if not os.path.isabs(name):
        return _error(
            ret, 'Specified file {0} is not an absolute path'.format(name))

    if isinstance(env, salt._compat.string_types):
        msg = (
            'Passing a salt environment should be done using \'saltenv\' not '
            '\'env\'. This warning will go away in Salt Boron and this '
            'will be the default and expected behaviour. Please update your '
            'state files.'
        )
        salt.utils.warn_until('Boron', msg)
        ret.setdefault('warnings', []).append(msg)
        # No need to set __env__ = env since that's done in the state machinery

    if os.path.isdir(name):
        ret['comment'] = 'Specified target {0} is a directory'.format(name)
        ret['result'] = False
        return ret

    if context is None:
        context = {}
    elif not isinstance(context, dict):
        return _error(
            ret, 'Context must be formed as a dict')

    if contents and contents_pillar:
        return _error(
            ret, 'Only one of contents and contents_pillar is permitted')

    # If contents_pillar was used, get the pillar data
    if contents_pillar:
        contents = __salt__['pillar.get'](contents_pillar)
        # Make sure file ends in newline
        if not contents.endswith('\n'):
            contents += '\n'

    if not replace and os.path.exists(name):
        # Check and set the permissions if necessary
        ret, _ = __salt__['file.check_perms'](name, ret, user, group, mode)
        if __opts__['test']:
            ret['comment'] = 'File {0} not updated'.format(name)
        elif not ret['changes'] and ret['result']:
            ret['comment'] = ('File {0} exists with proper permissions. '
                              'No changes made.'.format(name))
        return ret

    if name in _ACCUMULATORS:
        if not context:
            context = {}
        context['accumulator'] = _ACCUMULATORS[name]

    try:
        if __opts__['test']:
            ret['result'], ret['comment'] = __salt__['file.check_managed'](
                name,
                source,
                source_hash,
                user,
                group,
                mode,
                template,
                context,
                defaults,
                __env__,
                contents,
                **kwargs
            )
            return ret

        # If the source is a list then find which file exists
        source, source_hash = __salt__['file.source_list'](
            source,
            source_hash,
            __env__
        )
    except CommandExecutionError as exc:
        ret['result'] = False
        ret['comment'] = 'Unable to manage file: {0}'.format(exc)
        return ret

    # Gather the source file from the server
    try:
        sfn, source_sum, comment_ = __salt__['file.get_managed'](
            name,
            template,
            source,
            source_hash,
            user,
            group,
            mode,
            __env__,
            context,
            defaults,
            **kwargs
        )
    except Exception as exc:
        ret['changes'] = {}
        log.debug(traceback.format_exc())
        return _error(ret, 'Unable to manage file: {0}'.format(exc))

    if comment_ and contents is None:
        return _error(ret, comment_)
    else:
        try:
            return __salt__['file.manage_file'](
                name,
                sfn,
                ret,
                source,
                source_sum,
                user,
                group,
                mode,
                __env__,
                backup,
                makedirs,
                template,
                show_diff,
                contents,
                dir_mode)
        except Exception as exc:
            ret['changes'] = {}
            log.debug(traceback.format_exc())
            return _error(ret, 'Unable to manage file: {0}'.format(exc))

###
### Sensu specific internal functions
###

def _default_param(arglist, param_name, pillar = {}, default = None):
    if param_name in arglist:
        return
    if param_name in pillar:
        arglist[param_name] = pillar[param_name]
        return
    arglist[param_name] = default

def _sensucheck(name, **kwargs):

    # If the pillar kwarg is provided then use this to override the defaults
    # but ensure it is removed to avoid confusing things downstream
    pillar = {}
    if 'pillar' in kwargs:
        pillar = kwargs.pop('pillar')

    log.warning("Pillar is set to {}".format(repr(pillar)))
    _default_param(kwargs, 'command',     pillar, 'mycommand')
    _default_param(kwargs, 'handlers',    pillar, [ 'default' ])
    _default_param(kwargs, 'interval',    pillar, 60)
    _default_param(kwargs, 'occurrences', pillar, 1)
    _default_param(kwargs, 'playbook',    pillar, None)
    _default_param(kwargs, 'standalone',  pillar, False)
    _default_param(kwargs, 'subscribers', pillar, [ 'all' ])
  
    kwargs['source'] = 'salt://sensu/templates/state_checks.json'
    kwargs['template'] = 'jinja'
    kwargs['checkname'] = name
    pathname = "/etc/sensu/conf.d/checks/{}".format(name)
    return _managed(pathname, **kwargs)


###
### Sensu usable states
###

def check(name, **kwargs):
    return _sensucheck(name, **kwargs)


def check_procs(name, critical_under=1, **kwargs):
    pattern = kwargs.get('pattern',name)
    kwargs['command'] = "/etc/sensu/community/plugins/processes/check-procs.rb -p {} -C {}".format(pattern, critical_under)
    
    if 'critical_over' in kwargs:
        kwargs['command'] += " -c {}".format(kwargs['critical_over'])

    return _sensucheck(name, **kwargs)


def check_graphite(name, metric_name, params = '', desc = '', **kwargs):
    kwargs['command'] = "/etc/sensu/plugins/graphite-data.rb -t {} -n '{}' {}".format(metric_name, desc, params)
    
    return _sensucheck(name, **kwargs)


def check_by_name(name, **kwargs):
    ret = {'changes': {},
           'comment': '',
           'name': name,
           'result': True}
    pillar = kwargs.get('pillar', {})
    if pillar == {}:
        return _error(ret, "'pillar' not present") 
    type = pillar.get('type', None)
    if type == None:
        return _error(ret, "'type' not specified in pillar") 
    if type not in [ 'check_graphite', 'check', 'check_procs' ]:
        return _error(ret, "Unknown 'type' specified in pillar") 

    fn = globals()[type]
    
    return fn(name, **kwargs)


def handler(name, **kwargs):
    kwargs['source'] = 'salt://sensu/templates/state_handlers.json'
    kwargs['template'] = 'jinja'
    kwargs['handlername'] = name
    pathname = "/var/tmp/{}".format(name)
    return _managed(pathname, **kwargs)


